### PR TITLE
Implement gateway service

### PR DESCRIPTION
**Pull request by bennerv**
Implement gateway service (aka egress lockdown) [9618624](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9618624/).

Work still needed:
- [x] firewalling
- [x] unit tests
- [x] metrics
  - #connections (total)
  - #connections (new per second)
  - #goroutines
  - last successful change feed update
  - last successful db token refresh
  - CPU/Memory/Network on VMSS

Expected production impact:
* Big changes to the RP deployment because now we deploy gateway VMs, vnets, etc.
* Nearly all cluster changes are behind feature flag, so minimal impact to customers with this PR.
* PROD clusters in dev subscription + all INT clusters + PROD nightly E2E clusters will be gateway enabled.
* For all newly created clusters, regardless of feature flag, the RP will configure the cluster subnets as it needs to (no longer validates the ACR service endpoint or DisablePrivateLinkServiceNetworkPolicy flag but sets this as it wants to at cluster creation time).

Deployment plan:
* RP-Config update needed before merge - check the private gatewayDomains population tool, and see `git diff -w master pkg/deploy/config.go` for an indicator of what needs to be done
  * gatewayResourceGroupName: generate sensibly on a per-region basis
  * disableCosmosDBFirewall: omit in INT/PROD for obvious reasons
  * extraGatewayKeyvaultAccessPolicies: same as extraDBTokenKeyvaultAccessPolicies
  * gatewayDomains: use population tool
  * gatewayFeatures: [] in INT and PROD
  * gatewayMdsdConfigVersion: same as rpMdsdConfigVersion
  * gatewayStorageAccountDomain: generate sensibly on a per-region basis
  * gatewayVmssCapacity: omit in INT/PROD, will default to 3
* as the gateway is deployed, its gateway-mdsd (mdm, mdsd) certs will need to be generated at the same time (@mjudeikis or @jewzaam will need to help with this)
* documentation updates needed:
  * announcing feature
  * indicating that the customer no longer has to set --disable-private-link-service-network-policies true or --service-endpoints Microsoft.ContainerRegistry on their vnet (see https://github.com/MicrosoftDocs/azure-docs/pull/74910 )
* add preaggregates and dashboards for metrics (gateway component and Go metrics for all components)
* dbtoken aad application per tenant

    ```	
    "identifierUris": [
		"https://dbtoken.aro.azure.com"
	],
    ```

Follow-ups:
* Make Gateway VMs at least Standard_D4s_v3 and set EnableAcceleratedNetworking
* Use rpLBAlert() to define Azure monitoring alerts around the readiness of the gateway ILB
* On gateway upgrade, wait some time for old connections to die
* Detailed connection metrics
* E2E test coverage

### CODE REVIEW

**Comment by bennerv** [comment]
See #1368 for past comments.  Opened this as I can't push to Jim's branch in his fork.

**Comment by bennerv** [comment]
Putting a "hold" label on it as I want to get the PRs opened for RP-Config changes for Int/Prod.  Will take off once I get there.

**Comment by bennerv** [comment]
Relies on #1616 and #1617 to build properly.

**Comment by bennerv** [comment]
Holding (and not rebasing) until #1616 is merged.  Should also wait for https://github.com/mjudeikis/installer/pull/4 to be merged in first and then we can vendor MJ's installer instead of mine.

File: pkg/deploy/generator/templates_rp.go [hunk 1]
```diff
@@ -84,6 +87,8 @@ func (g *generator) rpTemplate() *arm.Template {
 			"armClientId",
 			"billingServicePrincipalId",
 			"billingE2EStorageAccountId",
+			"extraCosmosDBIPs",
```
Comment (troy0820) [inline pkg/deploy/generator/templates_rp.go:90]: Is there any reason to add this back? 
https://github.com/Azure/ARO-RP/pull/1561#issue-677937338
```suggestion
			
```

**Review by troy0820 — DISMISSED** [review]
This PR is huge.  I understand that you inherited this work but smaller changes would have been easier to review.  _(even doing this by commits it's a lot)_

I'm not done looking at all of this but the changes requested so far are for the `extraCosmosDBIPs` that were removed in https://github.com/Azure/ARO-RP/pull/1561#issue-677937338.  

I do see the TODOs there for the duplication of the deploy pkg that probably can reduced _(maybe in a subsequent PR?)_

I wasn't sure if the diagram of this new architecture would be included as well.

File: pkg/deploy/generator/templates_rp.go (hunk 1 above)
Comment (bennerv) [reply pkg/deploy/generator/templates_rp.go]: Fixed - most likely a merge change introduced on rebase.  Thank you!

File: pkg/deploy/generator/resources_gateway.go [hunk 2]
```diff
@@ -0,0 +1,715 @@
...
+[Install]
+WantedBy=multi-user.target
+EOF
+
+cat >/etc/sysconfig/aro-gateway <<EOF
+ACR_RESOURCE_ID='$ACRRESOURCEID'
+DATABASE_ACCOUNT_NAME='$DATABASEACCOUNTNAME'
+AZURE_DBTOKEN_CLIENT_ID='$DBTOKENCLIENTID'
+DBTOKEN_URL='$DBTOKENURL'
+MDM_ACCOUNT=AzureRedHatOpenShiftRP
```
Comment (bennerv) [inline pkg/deploy/generator/resources_gateway.go:372]: Needs to be cloudaware.

File: pkg/deploy/generator/resources_gateway.go [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+[Unit]
+After=network-online.target
+EOF
+
+cat >/etc/default/mdsd <<EOF
+MDSD_ROLE_PREFIX=/var/run/mdsd/default
+MDSD_OPTIONS="-A -d -r \$MDSD_ROLE_PREFIX"
+
+export MONITORING_GCS_ENVIRONMENT='$MDSDENVIRONMENT'
+export MONITORING_GCS_ACCOUNT=ARORPLogs
```
Comment (bennerv) [inline pkg/deploy/generator/resources_gateway.go:511]: Cloud aware

File: pkg/deploy/generator/resources_gateway.go [hunk 4]
```diff
@@ -0,0 +1,718 @@
...
+elif [ "\$COMPONENT" = "mdsd" ]; then
+  CURRENT_CERT_FILE="/var/lib/waagent/Microsoft.Azure.KeyVault.Store/mdsd.pem"
+else
+  echo Invalid usage && exit 1
+fi
+
+SECRET_NAME="gwy-\${COMPONENT}"
+NEW_CERT_FILE="\$TEMP_DIR/\$COMPONENT.pem"
+for attempt in {1..5}; do
+  az keyvault secret download --file \$NEW_CERT_FILE --id "https://$KEYVAULTPREFIX-gwy.vault.azure.net/secrets/\$SECRET_NAME" && break
```
Comment (bennerv) [inline pkg/deploy/generator/resources_gateway.go:477]: vault url is not cloud aware

**Comment by bennerv** [comment]
@mjudeikis @m1kola  should be good to go.  E2e passed in prod & int first redeploy after rebase.  

Had @jewzaam  try to create a cluster and it failed on workers because of a quota issue, but would have made it otherwise as the masters were all successful, etc.  

Think we can merge this whenever you guys are ready.

**Review by jewzaam — APPROVED** [review]
Testing in canary done, all green.  Verified cluster created w/o gateway works fine and when patched to turn on gateway operates as expected.  :shipit:

**Review by jewzaam — APPROVED** [review]
Updated to put "InProgress" feature back to forcing gateway enabled.  Ideally we'd run PUCM in E2E but we don't run actions in tests because of cert signing issues.  So we can't do that.  With this on we'll test gateway on every rollout.  Dev and CI will still run w/ gateway off (development mode).

**Pull request merged**